File: Assets/_Project/Scripts/Player/PlaneController.cs
using UnityEngine;

/// <summary>
/// Arcade-style plane controller for a 3D flight prototype.
/// Prioritizes readable controls and tunable handling over realism.
/// </summary>
[RequireComponent(typeof(Rigidbody))]
public class PlaneController : MonoBehaviour
{
    [Header("Config")]
    [SerializeField] private PlaneConfig config;

    [Header("Runtime State")]
    [SerializeField] private float currentSpeed;
    [SerializeField] private float currentBoost;
    [SerializeField] private float currentHealth;

    [Header("Debug")]
    [SerializeField] private bool lockCursorOnPlay;

    private Rigidbody rb;

    private float throttleInput;
    private float pitchInput;
    private float yawInput;
    private float rollInput;
    private bool boostInput;

    public float CurrentSpeed => currentSpeed;
    public float CurrentBoost => currentBoost;
    public float CurrentHealth => currentHealth;
    public float Altitude => transform.position.y;

    public float BoostNormalized
    {
        get
        {
            if (config == null || config.maxBoost <= 0f)
            {
                return 0f;
            }

            return currentBoost / config.maxBoost;
        }
    }

    public float HealthNormalized
    {
        get
        {
            if (config == null || config.maxHealth <= 0f)
            {
                return 0f;
            }

            return currentHealth / config.maxHealth;
        }
    }

    public bool IsBoosting => boostInput && currentBoost > 0f;

    public void SetConfig(PlaneConfig planeConfig)
    {
        config = planeConfig;
    }

    private void Awake()
    {
        rb = GetComponent<Rigidbody>();
    }

    private void Start()
    {
        if (config == null)
        {
            Debug.LogError("PlaneController: PlaneConfig is not assigned.");
            enabled = false;
            return;
        }

        rb.useGravity = false;
        rb.drag = 1f;
        rb.angularDrag = 2f;
        rb.interpolation = RigidbodyInterpolation.Interpolate;

        currentSpeed = Mathf.Clamp(config.startSpeed, config.minSpeed, config.maxSpeed);
        currentBoost = config.maxBoost;
        currentHealth = config.maxHealth;

        if (lockCursorOnPlay)
        {
            Cursor.lockState = CursorLockMode.Locked;
            Cursor.visible = false;
        }
    }

    private void Update()
    {
        ReadInput();
        UpdateSpeed();
        UpdateBoost();
    }

    private void FixedUpdate()
    {
        if (config == null)
        {
            return;
        }

        ApplyRotation();
        ApplyMovement();
    }

    private void ReadInput()
    {
        throttleInput = Input.GetAxis("Vertical");
        yawInput = Input.GetAxis("Horizontal");

        pitchInput = 0f;
        if (Input.GetKey(KeyCode.UpArrow))
        {
            pitchInput = 1f;
        }
        else if (Input.GetKey(KeyCode.DownArrow))
        {
            pitchInput = -1f;
        }

        rollInput = 0f;
        if (Input.GetKey(KeyCode.Q))
        {
            rollInput = 1f;
        }
        else if (Input.GetKey(KeyCode.E))
        {
            rollInput = -1f;
        }

        boostInput = Input.GetKey(KeyCode.LeftShift);
    }

    private void UpdateSpeed()
    {
        if (throttleInput > 0f)
        {
            currentSpeed += config.acceleration * throttleInput * Time.deltaTime;
        }
        else if (throttleInput < 0f)
        {
            currentSpeed += config.deceleration * throttleInput * Time.deltaTime;
        }

        currentSpeed = Mathf.Clamp(currentSpeed, config.minSpeed, config.maxSpeed);
    }

    private void UpdateBoost()
    {
        if (IsBoosting)
        {
            currentBoost -= config.boostDrainPerSecond * Time.deltaTime;
            currentBoost = Mathf.Max(currentBoost, 0f);
        }
        else
        {
            currentBoost += config.boostRechargePerSecond * Time.deltaTime;
            currentBoost = Mathf.Min(currentBoost, config.maxBoost);
        }
    }

    private void ApplyRotation()
    {
        float dt = Time.fixedDeltaTime;
        Quaternion targetRotation = rb.rotation;

        Quaternion pitchRotation = Quaternion.AngleAxis(pitchInput * config.pitchSpeed * dt, transform.right);
        Quaternion yawRotation = Quaternion.AngleAxis(yawInput * config.yawSpeed * dt, transform.up);
        Quaternion rollRotation = Quaternion.AngleAxis(-rollInput * config.rollSpeed * dt, transform.forward);

        targetRotation *= yawRotation * pitchRotation * rollRotation;

        Vector3 localEulerAngles = transform.localEulerAngles;
        float zAngle = localEulerAngles.z;

        if (zAngle > 180f)
        {
            zAngle -= 360f;
        }

        if (Mathf.Abs(rollInput) <= 0.01f)
        {
            float levelAdjustment = -zAngle * config.autoLevelStrength * dt;
            Quaternion autoLevelRotation = Quaternion.AngleAxis(levelAdjustment, transform.forward);
            targetRotation *= autoLevelRotation;
        }

        rb.MoveRotation(targetRotation);
    }

    private void ApplyMovement()
    {
        float appliedSpeed = currentSpeed;

        if (IsBoosting)
        {
            appliedSpeed *= config.boostMultiplier;
        }

        rb.velocity = transform.forward * appliedSpeed;
    }

    private void OnCollisionEnter(Collision collision)
    {
        if (config == null)
        {
            return;
        }

        TakeDamage(config.collisionDamage);
        currentSpeed *= 0.6f;
    }

    public void TakeDamage(float amount)
    {
        currentHealth -= amount;
        currentHealth = Mathf.Max(currentHealth, 0f);
    }

    public bool IsDestroyed()
    {
        return currentHealth <= 0f;
    }

    public void RestoreFullState()
    {
        if (config == null)
        {
            return;
        }

        currentSpeed = Mathf.Clamp(config.startSpeed, config.minSpeed, config.maxSpeed);
        currentBoost = config.maxBoost;
        currentHealth = config.maxHealth;
    }

    public void StopMotion()
    {
        rb.velocity = Vector3.zero;
        rb.angularVelocity = Vector3.zero;
    }
}

File: src/game/player/PlaneController.js
import { Quaternion, Vector3 } from "three";

const WORLD_UP = new Vector3(0, 1, 0);
const LOCAL_FORWARD = new Vector3(0, 0, 1);
const LOCAL_UP = new Vector3(0, 1, 0);
const TMP_FORWARD = new Vector3();
const TMP_UP = new Vector3();
const TMP_HORIZON_UP = new Vector3();
const TMP_CROSS = new Vector3();

export class PlaneController {
  constructor(object3d, config) {
    this.object3d = object3d;
    this.config = config;

    this.currentSpeed = config.startSpeed;
    this.currentBoost = config.maxBoost;
    this.currentHealth = config.maxHealth;
    this.boosting = false;
    this.spawnPosition = new Vector3();
    this.spawnRotation = new Quaternion();
  }

  setSpawn(position, quaternion) {
    this.spawnPosition.copy(position);
    this.spawnRotation.copy(quaternion);
    this.resetToSpawn();
  }

  update(dt, input) {
    this.updateSpeed(dt, input.throttle);
    this.updateBoost(dt, input.boost);
    this.applyRotation(dt, input);
    this.applyMovement(dt);
  }

  updateSpeed(dt, throttleInput) {
    if (throttleInput > 0) {
      this.currentSpeed += this.config.acceleration * throttleInput * dt;
    } else if (throttleInput < 0) {
      this.currentSpeed += this.config.deceleration * throttleInput * dt;
    }

    this.currentSpeed = clamp(this.currentSpeed, this.config.minSpeed, this.config.maxSpeed);
  }

  updateBoost(dt, boostInput) {
    this.boosting = boostInput && this.currentBoost > 0;

    if (this.boosting) {
      this.currentBoost -= this.config.boostDrainPerSecond * dt;
      this.currentBoost = Math.max(0, this.currentBoost);
    } else {
      this.currentBoost += this.config.boostRechargePerSecond * dt;
      this.currentBoost = Math.min(this.config.maxBoost, this.currentBoost);
    }
  }

  applyRotation(dt, input) {
    this.object3d.rotateY(degToRad(input.yaw * this.config.yawSpeed * dt));
    this.object3d.rotateX(degToRad(input.pitch * this.config.pitchSpeed * dt));
    this.object3d.rotateZ(degToRad(-input.roll * this.config.rollSpeed * dt));

    if (Math.abs(input.roll) <= 0.01) {
      const currentRoll = this.getRollAngle();
      this.object3d.rotateZ(-currentRoll * this.config.autoLevelStrength * dt);
    }
  }

  getRollAngle() {
    TMP_FORWARD.copy(LOCAL_FORWARD).applyQuaternion(this.object3d.quaternion).normalize();
    TMP_UP.copy(LOCAL_UP).applyQuaternion(this.object3d.quaternion).normalize();
    TMP_HORIZON_UP.copy(WORLD_UP).projectOnPlane(TMP_FORWARD);

    if (TMP_HORIZON_UP.lengthSq() < 0.0001) {
      return 0;
    }

    TMP_HORIZON_UP.normalize();
    TMP_CROSS.crossVectors(TMP_UP, TMP_HORIZON_UP);

    return Math.atan2(TMP_CROSS.dot(TMP_FORWARD), TMP_UP.dot(TMP_HORIZON_UP));
  }

  applyMovement(dt) {
    const speed = this.boosting ? this.currentSpeed * this.config.boostMultiplier : this.currentSpeed;
    TMP_FORWARD.copy(LOCAL_FORWARD).applyQuaternion(this.object3d.quaternion).normalize();
    this.object3d.position.addScaledVector(TMP_FORWARD, speed * dt);
  }

  applyCollisionPenalty(amount, collisionNormal) {
    this.takeDamage(amount);
    this.currentSpeed = Math.max(this.config.minSpeed, this.currentSpeed * 0.6);

    if (collisionNormal) {
      this.object3d.position.addScaledVector(collisionNormal, 4);
    }
  }

  takeDamage(amount) {
    this.currentHealth = Math.max(0, this.currentHealth - amount);
  }

  resetToSpawn() {
    this.object3d.position.copy(this.spawnPosition);
    this.object3d.quaternion.copy(this.spawnRotation);
    this.restoreFullState();
  }

  restoreFullState() {
    this.currentSpeed = clamp(this.config.startSpeed, this.config.minSpeed, this.config.maxSpeed);
    this.currentBoost = this.config.maxBoost;
    this.currentHealth = this.config.maxHealth;
    this.boosting = false;
  }

  get altitude() {
    return this.object3d.position.y;
  }

  get boostNormalized() {
    return this.config.maxBoost > 0 ? this.currentBoost / this.config.maxBoost : 0;
  }

  get healthNormalized() {
    return this.config.maxHealth > 0 ? this.currentHealth / this.config.maxHealth : 0;
  }

  get isDestroyed() {
    return this.currentHealth <= 0;
  }

  get position() {
    return this.object3d.position;
  }

  get quaternion() {
    return this.object3d.quaternion;
  }

  get collisionRadius() {
    return this.config.collisionRadius;
  }
}

function degToRad(value) {
  return value * (Math.PI / 180);
}

function clamp(value, min, max) {
  return Math.min(max, Math.max(min, value));
}
